tools/clang/unittests/HLSLExec/HlslExecTestUtils.cpp

```diff
@@ -232,133 +325,162 @@ static HRESULT enableAgilitySDK(HMODULE Runtime, UINT SDKVersion,
-      SDKPath.Append("\\");
+  if (!isExperimentalShadersEnabled())
+    return false;
+
+  HRESULT HR;
+  if (FAILED(HR = DeviceFactory->EnableExperimentalFeatures(
+                 1, &D3D12ExperimentalShaderModels, nullptr, nullptr))) {
+    LogWarningFmt(L"EnableExperimentalFeature(D3D12ExperimentalShaderModels) "
+                  L"failed: 0x%08x",
+                  HR);
+    return false;
   }
 
-  if (SDKPath.IsEmpty())
-    SDKPath = L".\\D3D12\\";
-
-  const bool MustFind = SDKVersion > 0;
-  if (SDKVersion <= 1) {
-    // lookup version from D3D12Core.dll
-    SDKVersion = getD3D12SDKVersion((LPCWSTR)SDKPath);
-    if (MustFind && SDKVersion == 0) {
-      hlsl_test::LogErrorFmt(L"Agility SDK not found in relative path: %s",
-                             (LPCWSTR)SDKPath);
-      return E_FAIL;
-    }
+  return true;
+}
+
+static CComPtr<ID3D12DeviceFactory>
+createDeviceFactorySDK(const AgilitySDKConfiguration &C) {
+  using namespace hlsl_test;
+
+  HRESULT HR;
+
+  CComPtr<ID3D12SDKConfiguration1> SDKConfig;
+  if (FAILED(HR = D3D12GetInterface(CLSID_D3D12SDKConfiguration,
+                                    IID_PPV_ARGS(&SDKConfig)))) {
+    LogCommentFmt(L"Failed to get ID3D12SDKConfiguration1 interface: 0x%08x",
+                  HR);
+    return nullptr;
   }
 
-  // Not found, not asked for.
-  if (SDKVersion == 0)
-    return S_FALSE;
-
-  HRESULT HR = enableAgilitySDK(Runtime, SDKVersion, (LPCWSTR)SDKPath);
-  if (FAILED(HR)) {
-    // If SDKVersion provided, fail if not successful.
-    // 1 means we should find it, and fill in the version automatically.
-    if (MustFind) {
-      hlsl_test::LogErrorFmt(
-          L"Failed to set Agility SDK version %d at path: %s", SDKVersion,
-          (LPCWSTR)SDKPath);
-      return HR;
-    }
-    return S_FALSE;
+  CComPtr<ID3D12DeviceFactory> DeviceFactory;
+  if (FAILED(
+          HR = SDKConfig->CreateDeviceFactory(C.SDKVersion, CW2A(C.SDKPath),
+                                              IID_PPV_ARGS(&DeviceFactory)))) {
+    LogCommentFmt(L"CreateDeviceFactory(%d, '%s', ...) failed: 0x%08x",
+                  C.SDKVersion, static_cast<const wchar_t *>(C.SDKPath), HR);
+    return nullptr;
   }
-  if (HR == S_OK)
-    hlsl_test::LogCommentFmt(L"Agility SDK version set to: %d", SDKVersion);
 
-  return HR;
+  LogCommentFmt(L"Using DeviceFactory for SDKVersion %d, SDKPath %s",
+                C.SDKVersion, static_cast<const wchar_t *>(C.SDKPath));
+
+  if (enableExperimentalMode(DeviceFactory))
+    LogCommentFmt(L"Experimental mode enabled.");
+  else
+    LogCommentFmt(L"Experimental mode not enabled.");
+
+  return DeviceFactory;
 }
 
-HRESULT enableExperimentalMode(HMODULE Runtime) {
-#ifdef _FORCE_EXPERIMENTAL_SHADERS
-  bool ExperimentalShaderModels = true;
-#else
-  bool ExperimentalShaderModels =
-      hlsl_test::GetTestParamBool(L"ExperimentalShaders");
-#endif // _FORCE_EXPERIMENTAL_SHADERS
+std::optional<D3D12SDK> D3D12SDK::create() {
```

Comment:
This seems to be always returning an `D3D12SDK` instance, `std::optional` is not necessary.

Reply:
good point!